Terminal.Gui/Views/ComboBox.cs

```diff
@@ -258,115 +275,136 @@ public override bool ProcessKey(KeyEvent e)
-			this.text = search.Text = text;
+			this.text = search.Text = text.ToString();
 			search.CursorPosition = 0;
 			search.TextChanged += Search_Changed;
 		}
 
+		private void Selected ()
+		{
+			if (listview.Source.Count == 0 || searchset.Count == 0) {
+				text = "";
+				return;
+			}
+
+			SetValue (searchset [listview.SelectedItem]);
+			search.CursorPosition = search.Text.RuneCount;
+			Search_Changed (search.Text);
+			OnSelectedChanged ();
+			Reset (keepSearchText: true);
+		}
+
 		/// <summary>
 		/// Reset to full original list
 		/// </summary>
-		private void Reset()
+		private void Reset (bool keepSearchText = false)
 		{
-			search.Text = text = "";
-			OnSelectedChanged();
+			if (!keepSearchText) {
+				search.Text = text = "";
+			}
 
 			ResetSearchSet ();
 
-			listview.SetSource(searchset);
+			listview.SetSource (searchset);
 			listview.Height = CalculatetHeight ();
 
-			this.SetFocus(search);
+			this.SetFocus (search);
 		}
 
-		private void ResetSearchSet()
+		private void ResetSearchSet (bool noCopy = false)
 		{
-			if (autoHide) {
-				if (searchset == null)
-					searchset = new List<string> ();
-				else
-					searchset.Clear ();
-			} else
-				searchset = source.ToList ();
+			if (searchset == null) {
+				searchset = new List<object> ();
+			} else { 
+				searchset.Clear ();
+			}
+
+			if (autoHide || noCopy)
+				return;
+
+			// force deep copy
+			foreach (var item in Source.ToList ()) {
+				searchset.Add (item);
+			}
 		}
 
 		private void Search_Changed (ustring text)
 		{
-			if (source == null) // Object initialization
+			if (source == null) { // Object initialization		
 				return;
+			}
 
-			if (string.IsNullOrEmpty (search.Text.ToString ()))
+			if (ustring.IsNullOrEmpty (search.Text)) {
 				ResetSearchSet ();
-			else
-				searchset = source.ToList().Cast<string>().Where (x => x.StartsWith (search.Text.ToString (), StringComparison.CurrentCultureIgnoreCase)).ToList();
+			} else {
+				ResetSearchSet (noCopy: true);
 
-			listview.SetSource (searchset);
-			listview.Height = CalculatetHeight ();
+				foreach (var item in source.ToList ()) { // Iterate to preserver object type and force deep copy
+					if (item.ToString().StartsWith (search.Text.ToString(), StringComparison.CurrentCultureIgnoreCase)) { 
+						searchset.Add (item);
+					}
+				}
+			}
 
-			listview.Redraw (new Rect (0, 0, width, height)); // for any view behind this
-			this.SuperView?.BringSubviewToFront (this);
+			ShowList ();
 		}
 
 		/// <summary>
-		/// Internal height of dynamic search list
+		/// Show the search list
 		/// </summary>
-		/// <returns></returns>
-		private int CalculatetHeight ()
+		/// 
+		/// Consider making public
+		private void ShowList ()
 		{
-			return Math.Min (height, searchset.Count);
+			listview.SetSource (searchset);
+			listview.Height = CalculatetHeight ();
+			listview.Redraw (new Rect (0, 0, Bounds.Width, Bounds.Height)); // Ensure list shrinks in Dialog
```

Comment:
Generally, I view it is a danger sign when code calls `Redraw` directly vs. `SetNeedsDisplay`. Are you sure you need to synchronously force a paint here? I don't fully understand the way this control works, so it may be needed. Thanks.

Reply:
He need it in the `Dialog`, but he could use the `Clear `method instead.

Reply:
I changed it to Clear() and it achieves the same effect. Thanks @BDisp 

*pushed*